utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py

```diff
@@ -0,0 +1,104 @@
+from typing import Optional, Dict
+import re
+import os
+
+# This allows users to refer to the value of the resolved variable in conditional statements e.g. 
+# "${VALUE?'!!_postfix'||null}"
+SELF_REFERENCE_MARKER = "!!"
+
+
+def perform_config_substitution(
+    input_string: str, replacements: Optional[Dict[str, str]] = None
+) -> str:
+    """
+    Perform case-insensitive, bash-style variable substitution on the input string with extended functionality.
+
+    Note that in all cases, any single quotes ' will be replaced with double quotes in the output ".
+
+    Please also note the special case of a conditional statement immediately surrounded by quotes. 
+    In this case the double quotes are stripped. But you can reintroduce by using single quotes. This 
+    allows the following style  "${VALUE?'!!'||null}" which remains valid in the JSON document.
+
+    The function supports the following patterns (in order of precedence):
+    1. "${VARIABLE_NAME?true_value||false_value}" - Quoted conditional substitution with else value
+    2. "${VARIABLE_NAME?true_value}" - Quoted conditional substitution without else value
+    3. ${VARIABLE_NAME?true_value||false_value} - Unquoted conditional substitution with else value
+    4. ${VARIABLE_NAME?true_value} - Unquoted conditional substitution without else value
+    5. ${VARIABLE_NAME:-default_value} - Substitution with default value
+    6. ${VARIABLE_NAME} - Basic substitution
+
+    Args:
+        input_string (str): The input string containing variables to be substituted.
```

Comment:
is this input_string, the string of one of the json config files?

Reply:
could be any config file yep

Reply:
will update